### PR TITLE
XML Converter

### PR DESCRIPTION
**Pull request by thbaan**
## Description

Following details are still missing:
 
- [x] Add support for declaring XML hierarchy for each property in a configuration
- [x] Add support for nullable properties when no matching elements in XML document exists
- [x] Add optional Translator Func to be used to translate incoming values from EDI to domain specific values
- [x] Removal of XmlDeserializer and fully switch to XmlConverter (including Dependency Injection setup)
- [x] Add AssertValidConfiguration (are all properties mapped?) and invoke from Unit Test

Potential improvements for later implementation:

 - Cache Compiled Lambda across requests
 - XmlTextReader (Fast Forward Reading for better support of reading large XML documents)
 - Add support for handling of complex types as properties in a record (AddSubType)

## References

* #0000

### CODE REVIEW

File: source/Energinet.DataHub.MeteringPoints.Infrastructure/EDI/XmlConverter/XmlMapper.cs [hunk 1]
```diff
@@ -0,0 +1,147 @@
...
+            }
+
+            if (element == null)
+            {
+                throw new Exception($"Element not found: {name}");
+            }
+
+            if (element.HasElements)
+            {
+                return element.Elements().First();
```
[the remark this answers is not in the document]
Comment (thbaan) [reply source/Energinet.DataHub.MeteringPoints.Infrastructure/EDI/XmlConverter/XmlMapper.cs]: This is now resolved in latest update of this PR. It´s now possible to declare a specific XML hiearchy for each property.

File: source/Energinet.DataHub.MeteringPoints.Infrastructure/EDI/XmlConverter/XmlMapper.cs [hunk 2]
```diff
@@ -0,0 +1,147 @@
...
+                var args = properties.Select(x =>
+                {
+                    if (x.Key == nameof(Application.CreateMeteringPoint.InstallationLocationAddress))
+                    {
+                        return new Address(); // TODO: mapper must be able to handle complex sub types
+                    }
+
+                    if (x.Value is null)
+                    {
+                        throw new ArgumentNullException($"Missing map for property with name: {x.Key}");
```
[the remark this answers is not in the document]
Comment (thbaan) [reply source/Energinet.DataHub.MeteringPoints.Infrastructure/EDI/XmlConverter/XmlMapper.cs]: Actually this is not an evaluation of the XML element value itself. This code checks if theres a mapping for the specific property.

File: source/Energinet.DataHub.MeteringPoints.Infrastructure/EDI/XmlConverter/ConverterMapperConfigurations.cs [hunk 3]
```diff
@@ -0,0 +1,67 @@
...
+            }
+        }
+
+        private static List<XmlMappingConfigurationBase> GetAllConfigurations()
+        {
+            return typeof(XmlMappingConfigurationBase).Assembly.GetTypes()
+                .Where(t => t.IsSubclassOf(typeof(XmlMappingConfigurationBase)) && !t.IsAbstract)
+                .Select(Activator.CreateInstance)
+                .Cast<XmlMappingConfigurationBase>()
+                .ToList();
```
Comment (Sondergaard) [inline source/Energinet.DataHub.MeteringPoints.Infrastructure/EDI/XmlConverter/ConverterMapperConfigurations.cs:57]: This implies that all implementations of XmlMappingConfigurationBase is within the ...DataHub.MeteringPoints.Infrastructure assembly.

If this is meant to be a more general solution, then the discovery of assemblies should be more "open"

Comment (thbaan) [reply source/Energinet.DataHub.MeteringPoints.Infrastructure/EDI/XmlConverter/ConverterMapperConfigurations.cs:57]: Yep - I´m aware of that. Thanks for noticing!

File: source/Energinet.DataHub.MeteringPoints.Infrastructure/EDI/XmlConverter/XmlMapper.cs [hunk 4]
```diff
@@ -0,0 +1,126 @@
...
+            {
+                return valueTranslatorFunc != null ? valueTranslatorFunc(source ?? string.Empty) : source;
+            }
+
+            if (dest == typeof(bool))
+            {
+                return valueTranslatorFunc != null ? valueTranslatorFunc(source ?? string.Empty) : source;
+            }
+
+            return System.Convert.ChangeType(source, dest);
```
Comment (Sondergaard) [inline source/Energinet.DataHub.MeteringPoints.Infrastructure/EDI/XmlConverter/XmlMapper.cs:123]: What is the cost of this method call?

Is it possible to use the explicit methods on XElement?
https://docs.microsoft.com/en-us/dotnet/api/system.xml.linq.xelement.op_explicit?view=net-5.0

**Review by Sondergaard — COMMENTED** [review]
I have reviewed the code and found some Api design that is a bit 'meehhh'. It is about the need to invoke methods in a special order, if you don't want an exception thrown in your face.

The discovery of mappings would also need a rework if turned into a nuget package.

I can see this can work for simple types as in your example.
If this should be used in other domains where xml-elements have an iteration then it could be difficult to extend this library to suit that need.

**Review by cmgrd — APPROVED** [review]

**Review by erduna — APPROVED** [review]

**Pull request merged**